### DISCUSSION TITLE
Metrics for latency caused by each plugin for a request

### DISCUSSION THREAD

**Discussion by nishant95** [Kong/kong#10618] [Ideas and feature requests]
It seems that there isn't a straight forward way to attribute latency caused by each plugin for a request. Or any other way to benchmark performance of each plugin.

It would be beneficial if we can add metrics for this.

If this is already in place or is being discussed somewhere, please redirect and feel free to close this discussion.

**Comment by nishant95** [comment]
The answer [here](https://github.com/Kong/kong/discussions/8258#discussioncomment-1908909) requires one to change the code for the plugin. If this can be made available out of the box (so that it works for default plugins as well) then it would be a nice addition.

**Comment by ghost** [comment]
A request will go through stages such as rewrite, access, balancer, header_filter, and body_filter. Each plugin will hook some or all of these stages, and it will be quite complex to do such statistics and summary. In your scenario, are these metrics mainly used to observe which plugin is causing performance problems?

**Comment by nishant95** [comment]
@sabertobihwy 
Yes, It will help us debug request latencies as we would be able to differentiate between time taken by different plugins and time taken by rest of the process like kong processing/nginx processing/nginx queue wait times.

Also, we might want to monitor this in a prometheus dashboard.

**Comment by nishant95** [comment]
> Each plugin will hook some or all of these stages, and it will be quite complex to do such statistics and summary.

I think we should be able to instrument plugins iterator for this and add the recorded latencies to Kong ctx.
If you find this to be a nice addition then we can take a stab at it and draft a PR.

**Comment by ghost** [comment]
Usually, in the Kong field, people use flame graphs to troubleshoot performance issues.

**Comment by nishant95** [comment]
@sabertobihwy 
Profiling would not give visibility into the time spent by each request into each phase/plugin.

On a side-note, do we have any recommended way of profiling Kong and Kong plugins?
We are aware of some commercial solutions like Openresty XRay, but could not find something for Nginx and Lua other than the old system tap based library from openresty.

**Reply by ghost** [reply]
> On a side-note, do we have any recommended way of profiling Kong and Kong plugins?

I am not an official personnel. Based on my personal experience, perf and system tap have met most of the performance analysis scenarios.

> but could not find something for Nginx and Lua other than the old system tap based library from openresty.

 I don't know either. I'm still using them, such as [openresty-systemtap-toolkit](https://github.com/openresty/openresty-systemtap-toolkit), and they can still help me solve problems.

Just to add, I think relying on Kong's own plugin code to detect performance is not universal enough. It is similar to instrumentation or burying points. If the performance issue is not related to the plugin, then there is no solution. I am also looking for suitable base on eBPF performance observability tools for OpenResty and Kong, but have not found any suitable ones.
